source/dealii/dealii_hierarchy_helpers.cc

```diff
@@ -195,23 +202,25 @@ DealIIHierarchyHelpers<dim, VectorType>::build_restrictor(
               dealii::Vector<ScalarType> correction(n_elem);
               agglomerate_system_matrix.vmult(correction, delta_eig);
 
-              // We would like to fill the delta correction matrix but we can't
-              // because we don't know the sparsity pattern. So we accumulate
-              // all the values and then fill the matrix using the set()
-              // function.
-              for (unsigned int k = 0; k < n_elem; ++k)
-              {
-                local_copy_data.delta_correction_local_acc[std::make_pair(
-                    global_row, dof_indices_map[k])] += correction[k];
-              }
+              // Store the values the delta correction matrix is to be filled
+              // with.
+              local_copy_data.rows[j] = global_row;
+              std::transform(correction.begin(), correction.end(),
+                             local_copy_data.values_per_row[j].begin(),
+                             local_copy_data.values_per_row[j].begin(),
+                             std::plus<double>());
             }
           };
 
       auto copier = [&](const CopyData &local_copy_data) {
-        for (const auto &local_pair :
-             local_copy_data.delta_correction_local_acc)
+        for (unsigned int i = 0; i < local_copy_data.rows.size(); ++i)
         {
-          delta_correction_acc[local_pair.first] += local_pair.second;
+          // Since we didn't provide a sparsity pattern this "set" operation
+          // actually acts as an "add" operation until "compress" is called.
+          delta_correction_matrix.set(
+              local_copy_data.rows[i], local_copy_data.cols.size(),
+              local_copy_data.cols.data(),
+              local_copy_data.values_per_row[i].data(), true);
```

Comment:
Just out of curiosity, were the zero entries a significant part of the matrix?

Reply:
I didn't check so far.